File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-5.java
class Solution {
    public int findMin(int[] nums) {
        int n = nums.length;
        int left = 0;
        int right = n - 1;
        while (left <= right) {
            if (nums[left] <= nums[right]) return nums[left];
            int mid = (left + right) / 2;
            if (mid > 0 && nums[mid] < nums[mid - 1]) return nums[mid];
            if (mid < n - 1 && nums[mid] > nums[mid + 1]) return nums[mid + 1];
            if (nums[mid] >= nums[left]) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return -1;
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-5.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        unordered_map<char, int> counts;
        int left = 0;
        int maxFreq = 0;
        int res = 0;
        for (int right = 0; right < s.size(); right++) {
            counts[s[right]]++;
            maxFreq = max(counts[s[right]], maxFreq);
            while ((right - left + 1) - maxFreq > k) {
                counts[s[left]]--;
                left++;
            }
            res = max(right - left + 1, res);
        }
        return res;
    }
};
